### PR TITLE
Issue #145: Adds docker-compose for local development

### PR DESCRIPTION
**Pull request by SonOfLope**
ref #145  and https://github.com/ai-cfia/devops/issues/43

config for .env file can be found in vault under fertiscan-dev

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    container_name: frontend
+    restart: always
+    ports:
+      - "3000:3000"
+    env_file:
+      - .env
+    depends_on:
+      - postgres
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost:3000/health"]
```
Comment (k-allagbe) [inline docker-compose.yml:30]: In the frontend, we don't have code for this.

File: README.md [hunk 2]
```diff
@@ -67,6 +69,43 @@ and form generation using an [LLM](https://en.wikipedia.org/wiki/Large_language_
...
+    ```ini
+    BB_URL=bytebase_url
+    BB_SERVICE_ACCOUNT=your-bytebase-sa@service.bytebase.com
+    BB_SERVICE_KEY=your-bytebase-sa-key
+    BB_INSTANCE_ID=your-bytebase-instance-id
+    BB_DATABASE_ID=your-bytebase-database-id
+    ```
+
+    You can find their values in our vault under fertiscan-dev.
+
```
Comment (k-allagbe) [inline README.md:86]: There should be a mention that the `FERTISCAN_DB_URL` env var needs to point to the `postgres` service. In this particular case: `postgresql://postgres:postgres@postgres:5432/fertiscan`.

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      - fertiscan-network
+
+  frontend:
+    image: ghcr.io/ai-cfia/fertiscan-frontend:main
+    container_name: frontend
+    restart: always
+    ports:
+      - "3000:3000"
+    env_file:
+      - .env
```
Comment (k-allagbe) [inline docker-compose.yml:26]: Is it fine that we use the same env file for the services? I don't have an opinion on this.

File: docker-compose.yml (hunk 1 above)
Comment (SonOfLope) [reply docker-compose.yml]: I removed it thank you

File: README.md (hunk 2 above)
Comment (SonOfLope) [reply README.md:87]: I updated the readme

File: docker-compose.yml (hunk 3 above)
Comment (SonOfLope) [reply docker-compose.yml:26]: Yes. Variables from the frontend are different so they wont interfere. If needed the user can create a different file and update this line.

**Review by k-allagbe — COMMENTED** [review]
This is more a frontend issue I think, but because the way we handle the env vars in the frontend, no matter what .env we feed the compose file, the frontend requests on a /api endpoint. Example:
`frontend            |  HTTP  9/18/2024 4:35:12 PM 192.168.65.1 POST /api//analyze`

Probably due to this: https://github.com/ai-cfia/fertiscan-frontend/blob/main/.env.production

Any input on how we might handle this in the frontend would be great.

@Youllou, please also don't hesitate to chip in.

**Comment by SonOfLope** [comment]
> This is more a frontend issue I think, but because the way we handle the env vars in the frontend, no matter what .env we feed the compose file, the frontend requests on a /api endpoint. Example: `frontend | HTTP 9/18/2024 4:35:12 PM 192.168.65.1 POST /api//analyze`
> 
> Probably due to this: https://github.com/ai-cfia/fertiscan-frontend/blob/main/.env.production
> 
> Any input on how we might handle this in the frontend would be great.
> 
> @Youllou, please also don't hesitate to chip in.

should be fixed in https://github.com/ai-cfia/fertiscan-frontend/pull/244

We were building the image with 'api/' instead of '/api'

**Comment by k-allagbe** [comment]
> should be fixed in [ai-cfia/fertiscan-frontend#244](https://github.com/ai-cfia/fertiscan-frontend/pull/244)

It did fix it. But now I have another issue:
- After launching the compose, browse localhost:80
- import a fertilizer label image and submit
- observe that there is no result

Browser console logs:
image

502 being bad gateway, could this be related to nginx?

Docker compose logs:
```
backend             | 2024-09-19 02:22:03,318 - werkzeug - INFO - 127.0.0.1 - - [19/Sep/2024 02:22:03] "GET /health HTTP/1.1" 200 -
fertiscan-nginx     | 2024/09/19 02:22:05 [error] 9#9: *11 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.65.1, server: , request: "POST /api/analyze HTTP/1.1", upstream: "http://172.18.0.4:5001/analyze", host: "localhost", referrer: "http://localhost/"
fertiscan-nginx     | 192.168.65.1 - test-user-2 [19/Sep/2024:02:22:05 +0000] "POST /api/analyze HTTP/1.1" 502 157 "http://localhost/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:130.0) Gecko/20100101 Firefox/130.0"
backend             | 2024-09-19 02:22:33,386 - werkzeug - INFO - 127.0.0.1 - - [19/Sep/2024 02:22:33] "GET /health HTTP/1.1" 200 -
Gracefully stopping... (press Ctrl+C again to force)
```

Note that the backend receives nothing.

**Comment by SonOfLope** [comment]
> > should be fixed in [ai-cfia/fertiscan-frontend#244](https://github.com/ai-cfia/fertiscan-frontend/pull/244)
> 
> It did fix it. But now I have another issue:
> 
>     * After launching the compose, browse localhost:80
> 
>     * import a fertilizer label image and submit
> 
>     * observe that there is no result
> 
> 
> Browser console logs: image
> 
> 502 being bad gateway, could this be related to nginx?
> 
> Docker compose logs:
> 
> ```
> backend             | 2024-09-19 02:22:03,318 - werkzeug - INFO - 127.0.0.1 - - [19/Sep/2024 02:22:03] "GET /health HTTP/1.1" 200 -
> fertiscan-nginx     | 2024/09/19 02:22:05 [error] 9#9: *11 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.65.1, server: , request: "POST /api/analyze HTTP/1.1", upstream: "http://172.18.0.4:5001/analyze", host: "localhost", referrer: "http://localhost/"
> fertiscan-nginx     | 192.168.65.1 - test-user-2 [19/Sep/2024:02:22:05 +0000] "POST /api/analyze HTTP/1.1" 502 157 "http://localhost/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:130.0) Gecko/20100101 Firefox/130.0"
> backend             | 2024-09-19 02:22:33,386 - werkzeug - INFO - 127.0.0.1 - - [19/Sep/2024 02:22:33] "GET /health HTTP/1.1" 200 -
> Gracefully stopping... (press Ctrl+C again to force)
> ```
> 
> Note that the backend receives nothing.

As we spoke at the office, the issue was caused by the ports that were changed in the compose configuration. I added comments on the compose explaining how to handle port configuration.

**Review by k-allagbe — APPROVED** [review]
Tested and works as expected for me. Thank you.

**Pull request merged**